### PR TITLE
[AIRFLOW-4899] Fix BQ Hook - get_datasets_list uses pagination (#6780)

### PR DESCRIPTION
**Pull request by benjamingrenier**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4899
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by TobKed** [comment]
@benjamingrenier could you fix link to the Jira issue in the description please?

File: airflow/gcp/hooks/bigquery.py [hunk 1]
```diff
@@ -1798,17 +1798,22 @@ def get_dataset(self, dataset_id: str, project_id: Optional[str] = None) -> Dict
 
         return dataset_resource
 
-    def get_datasets_list(self, project_id: Optional[str] = None) -> List:
+    @CloudBaseHook.catch_http_exception
+    def get_datasets_list(self, project_id=None, max_results=None, all_datasets=False):
```
Comment (TobKed) [inline airflow/gcp/hooks/bigquery.py:1802]: Could you add type annotations, please? I think it is good idea to keep them :)

File: airflow/gcp/hooks/bigquery.py [hunk 2]
```diff
@@ -1833,16 +1838,26 @@ def get_datasets_list(self, project_id: Optional[str] = None) -> List:
...
+            optional_params['maxResults'] = max_results
 
-        except HttpError as err:
-            raise AirflowException(
-                'BigQuery job failed. Error was: {}'.format(err.content))
+        request = self.service.datasets().list(
+            projectId=dataset_project_id,
+            **optional_params)
+
+        datasets_list = []
```
Comment (TobKed) [inline airflow/gcp/hooks/bigquery.py:1852]: I think you could just use `datasets` here and add type. What do you think?

File: airflow/gcp/hooks/bigquery.py [hunk 3]
```diff
@@ -1833,16 +1838,26 @@ def get_datasets_list(self, project_id: Optional[str] = None) -> List:
...
+        optional_params = {'all': all_datasets}
+        if max_results:
+            optional_params['maxResults'] = max_results
 
-        except HttpError as err:
-            raise AirflowException(
-                'BigQuery job failed. Error was: {}'.format(err.content))
+        request = self.service.datasets().list(
+            projectId=dataset_project_id,
+            **optional_params)
```
Comment (TobKed) [inline airflow/gcp/hooks/bigquery.py:1850]: I wonder is this `optional_params` dictionary is required. I've checked how it is done here:  `airflow.providers.google.marketing_platform.hooks.campaign_manager.GoogleCampaignManagerHook.list_reports` and it is slightly cleaner. WDYT?

File: tests/gcp/hooks/test_bigquery.py [hunk 4]
```diff
@@ -795,18 +795,48 @@ def test_get_datasets_list(self):
...
-                               'datasets') as MockService:
-            MockService.return_value.list(
-                projectId=project_id).execute.return_value = expected_result
-            result = hook.BigQueryBaseCursor(
-                mocked, "test_create_empty_dataset").get_datasets_list(
-                project_id=project_id)
-            self.assertEqual(result, expected_result['datasets'])
+        mock_service = mock.Mock()
+        cursor = hook.BigQueryBaseCursor(mock_service, project_id)
+        mock_service.datasets.return_value.list.return_value.execute.return_value = {
```
Comment (TobKed) [inline tests/gcp/hooks/test_bigquery.py:803]: I think you could check here by using `assert_called_once_with` with what parameters list method was called.

**Review by TobKed — CHANGES_REQUESTED** [review]
Good job @benjamingrenier :) . I've left some comments

**Comment by Mofef** [comment]
@benjamingrenier I just ran into the same issue and found that you were apparently trying to get this fix merged since July 2019 (#5565). Would you be willing to pick this up again? Can I help?

**Comment by mik-laj** [comment]
@Mofef Can you create a new PR? I am happy to help with the review.

**Comment by Mofef** [comment]
Sure, I'd like to try that. Is it ok to start with forking from what @benjamingrenier already achieved and addressing the previous review comments?

**Pull request closed without merging**